submariner/pluggable-network-plugin-and-ovn.md

```diff
@@ -0,0 +1,300 @@
+## Add support for OVN with pluggable network-plugin support
+
+[submariner#778](https://github.com/submariner-io/submariner/issues/778)
+
+## Summary
+
+Today Submariner supports working with existing network-plugins which are based
+on the kube-proxy with iptables backend. Any network-plugin outside of this
+case won't work, OVN-Kubernetes is an example.
+
+This is a proposal to support pluggable network drivers, starting with OVN-Kubernetes.
+
+## Proposal
+
+Re-architect Submariner route-agent in a way such that different network-plugin implementations
+can be supported. Those drivers would be notified of different events like:
+
+* `Initialization` event
+    Called once during startup, to let the plugin configure any necessary settings on the
+    system once the route-agent is ready
+
+* `Stop` event
+    Called once during stop to let the network plugin cleanup anything necessary during
+    at stop if at all.
+
+* `Uninstall` event
+    Called once if the route agent is being removed from a node, being Submariner un-installed,
+    or the route-agent driver being switched, a full cleanup will be necessary in this case.
+
+* `Leader`/`Non-leader` events
+    Some parts of the drivers could need a single leader handling specific synchronization
+    work with the network plugin. Those events will be called for drivers that require
+    this type of behavior.
+
+    We may want to add an indication on the driver, to say if the driver needs leader election
+    or not. kube-proxy iptables does not need it, and leader election loads the K8s API.
+
+* `Transition to Non-Gateway` event
+    During Gateway to Non-Gateway events the route agent on the specific node
+    may need to perform specific cleanups
+
+* `Transition to Gateway` event
+    When a node becomes Gateway some settings may need to be adjusted on the node
+    by the driver.
```

Comment:
This sounds very specific to iptables but I'm not sure if it makes sense for other driver types..

Also if we model it more "micro-services" like then I'm not sure we need too many specific events as we could just have the different parts react to different states of resources in K8s.

Reply:
This part is also important for OVN, ovn will need to know which one is the active gateway, so it can re-assign the chassis-gateway for the router external port, receiving events for that it's important.

Reply:
AFAIU this will be required for most of the drivers as Gatewaynode requires different configuration compared to other nodes in the cluster.